items.c

```diff
@@ -612,3 +636,306 @@ void do_item_flush_expired(void) {
+            /* Attempt to hash item lock the "search" item. If locked, no
+             * other callers can incr the refcount
+             */
+            if ((hold_lock = item_trylock(hv)) == NULL) {
+                pthread_mutex_unlock(&cache_lock);
+                continue;
+            }
+            /* Now see if the item is refcount locked */
+            if (refcount_incr(&search->refcount) != 2) {
+                refcount_decr(&search->refcount);
+                if (hold_lock)
+                    item_trylock_unlock(hold_lock);
+                pthread_mutex_unlock(&cache_lock);
+                continue;
+            }
+
+            /* Frees the item or decrements the refcount. */
+            /* Interface for this could improve: do the free/decr here
+             * instead? */
+            item_crawler_evaluate(search, hv, i);
+
+            if (hold_lock)
+                item_trylock_unlock(hold_lock);
+            pthread_mutex_unlock(&cache_lock);
+
+            if (settings.lru_crawler_sleep)
+                usleep(settings.lru_crawler_sleep);
+        }
+    }
+    if (settings.verbose > 2)
+        fprintf(stderr, "LRU crawler thread sleeping\n");
+    STATS_LOCK();
+    stats.lru_crawler_running = false;
+    STATS_UNLOCK();
+    pthread_mutex_unlock(&lru_crawler_lock);
+    }
+    if (settings.verbose > 2)
+        fprintf(stderr, "LRU crawler thread stopping\n");
+
+    return NULL;
+}
+
+static pthread_t item_crawler_tid;
+
+int stop_item_crawler_thread(void) {
+    int ret;
+    pthread_mutex_lock(&lru_crawler_lock);
+    do_run_lru_crawler_thread = 0;
+    pthread_cond_signal(&lru_crawler_cond);
+    pthread_mutex_unlock(&lru_crawler_lock);
+    if ((ret = pthread_join(item_crawler_tid, NULL)) != 0) {
+        fprintf(stderr, "Failed to stop LRU crawler thread: %s\n", strerror(ret));
+        return -1;
+    }
+    settings.lru_crawler = false;
+    return 0;
+}
+
+int start_item_crawler_thread(void) {
+    int ret;
+
+    if (settings.lru_crawler)
+        return -1;
+    pthread_mutex_lock(&lru_crawler_lock);
+    do_run_lru_crawler_thread = 1;
+    settings.lru_crawler = true;
+    if ((ret = pthread_create(&item_crawler_tid, NULL,
+        item_crawler_thread, NULL)) != 0) {
+        fprintf(stderr, "Can't create LRU crawler thread: %s\n",
+            strerror(ret));
+        pthread_mutex_unlock(&lru_crawler_lock);
+        return -1;
+    }
+    pthread_mutex_unlock(&lru_crawler_lock);
+
+    return 0;
+}
+
+enum crawler_result_type lru_crawler_crawl(char *slabs) {
+    char *b = NULL;
+    uint32_t sid = 0;
+    uint8_t tocrawl[POWER_LARGEST];
+    if (pthread_mutex_trylock(&lru_crawler_lock) != 0) {
+        return CRAWLER_RUNNING;
+    }
+    pthread_mutex_lock(&cache_lock);
+
+    if (strcmp(slabs, "all") == 0) {
+        for (sid = 0; sid < LARGEST_ID; sid++) {
+            tocrawl[sid] = 1;
+        }
+    } else {
+        for (char *p = strtok_r(slabs, ",", &b);
+             p != NULL;
```

Comment:
wow. the strtok_r completely threw me off. I thought the for loop was terminated. :)